### PR TITLE
New C++ Api: Comprehensive guards for member functions of class Term.

### PR DESCRIPTION
**Pull request by aniemetz**

### CODE REVIEW

File: src/api/cvc4cpp.cpp [hunk 1]
```diff
@@ -2239,155 +2255,155 @@ Op Term::getOp() const
...
+  //////// all checks before this line
+  return isNullHelper();
+  ////////
+  CVC4_API_TRY_CATCH_END;
+}
 
 Term Term::getConstArrayBase() const
 {
   NodeManagerScope scope(d_solver->getNodeManager());
+  CVC4_API_TRY_CATCH_BEGIN;
```
Comment (ajreynol) [inline src/api/cvc4cpp.cpp:2274]: Is the NodeManagerScope above this line necessary?

File: test/unit/api/term_black.cpp [hunk 2]
```diff
@@ -657,10 +657,10 @@ TEST_F(TestApiBlackTerm, termChildren)
   Term f = d_solver.mkConst(fsort, "f");
   Term t2 = d_solver.mkTerm(APPLY_UF, f, two);
   // due to our higher-order view of terms, we treat f as a child of APPLY_UF
-  ASSERT_EQ(t2.getNumChildren(), 2);
-  ASSERT_EQ(t2[0], f);
-  ASSERT_EQ(t2[1], two);
-  ASSERT_THROW(tnull[0], CVC4ApiException);
+  //ASSERT_EQ(t2.getNumChildren(), 2);
```
Comment (ajreynol) [inline test/unit/api/term_black.cpp:660]: Did you mean to comment this out?

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/api/cvc4cpp.cpp (hunk 1 above)
Comment (aniemetz) [reply src/api/cvc4cpp.cpp:2274]: I don't know but I'm tempted to assume it is. Ideally all of the functions should open a scope, but as discussed online, that is a huge can of worms that needs to be fixed.

File: test/unit/api/term_black.cpp (hunk 2 above)
Comment (aniemetz) [reply test/unit/api/term_black.cpp:660]: I did not.

File: src/api/cvc4cpp.cpp (hunk 1 above)
Comment (ajreynol) [reply src/api/cvc4cpp.cpp:2274]: Ah I see, I actually though you were opening as part of one of these macros.  I guess it's fine to keep for now.

Comment (aniemetz) [reply src/api/cvc4cpp.cpp:2274]: Yeah, adding this to the macro is the next step, but my initial attempt failed, so I'm doing this incrementally. We're close, though, now, so I'll try to add it after the refactor is in.

File: src/api/cvc4cpp.cpp [hunk 3]
```diff
@@ -2126,99 +2120,121 @@ size_t Term::getNumChildren() const
...
   {
-    CVC4_API_CHECK(d_node->hasOperator())
-        << "Expected apply kind to have operator when accessing child of Term";
-    if (index == 0)
+    if(index == 0)
     {
-      // return the operator
-      return Term(d_solver, d_node->getOperator());
+    // return the operator
+    return Term(d_solver, d_node->getOperator());
```
Comment (ajreynol) [inline src/api/cvc4cpp.cpp:2142]: Remember to clang-format.

**Review by ajreynol — APPROVED** [review]
LGTM, remember to clang-format.

**Pull request merged**